File: logic-programs/control.pl
% Control Systems

:- op(1200, xfx, :+).

% measurements
'<urn:example:measurement1>'('<urn:example:input1>', [6, 11]).
'<urn:example:measurement1>'('<urn:example:disturbance2>', [45, 39]).
'<urn:example:measurement2>'('<urn:example:input2>', true).
'<urn:example:measurement3>'('<urn:example:input3>', 56967).
'<urn:example:measurement3>'('<urn:example:disturbance1>', 35766).
'<urn:example:measurement4>'('<urn:example:output2>', 24).

% observations
'<urn:example:observation1>'('<urn:example:state1>', 80).
'<urn:example:observation2>'('<urn:example:state2>', false).
'<urn:example:observation3>'('<urn:example:state3>', 22).

% targets
'<urn:example:target2>'('<urn:example:output2>', 29).

% rules
'<urn:example:control1>'('<urn:example:actuator1>', C) :-
    '<urn:example:measurement10>'('<urn:example:input1>', M1),
    '<urn:example:measurement2>'('<urn:example:input2>', true),
    '<urn:example:measurement3>'('<urn:example:disturbance1>', D1),
    C1 is M1*19.6,          % proportial part
    C2 is log(D1)/log(10),  % compensation part
    C is C1-C2.             % simple feedforward control

'<urn:example:control1>'('<urn:example:actuator2>', C) :-
    '<urn:example:observation3>'('<urn:example:state3>', P3),
    '<urn:example:measurement4>'('<urn:example:output2>', M4),
    '<urn:example:target2>'('<urn:example:output2>', T2),
    E is T2-M4,             % error
    D is P3-M4,             % differential error
    C1 is 5.8*E,            % proportial part
    N is 7.3/E,             % nonlinear factor
    C2 is N*D,              % nonlinear differential part
    C is C1+C2.             % PND feedback control

'<urn:example:measurement10>'(I, M) :-
    '<urn:example:measurement1>'(I, [M1, M2]),
    M1 < M2,
    M3 is M2-M1,
    M is sqrt(M3).

'<urn:example:measurement10>'(I, M1) :-
    '<urn:example:measurement1>'(I, [M1, M2]),
    M1 >= M2.

% query
true :+ '<urn:example:control1>'(_, _).

File: eyelet/input/control.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>
#include <string.h>

// ──────────────────────────────────────────────────────────────
// ❶ FACTS  (the stuff after the dots in Prolog)
// ──────────────────────────────────────────────────────────────

// measurements
typedef struct {
    int input1[2];
    int disturbance2[2];
} Measurement1;

typedef struct {
    bool input2;
} Measurement2;

typedef struct {
    int input3;
    int disturbance1;
} Measurement3;

typedef struct {
    int output2;
} Measurement4;

// observations
typedef struct {
    int state1;
} Observation1;

typedef struct {
    bool state2;
} Observation2;

typedef struct {
    int state3;
} Observation3;

// targets
typedef struct {
    int output2;
} Target2;

// Initialize facts
Measurement1 measurement1 = {
    .input1 = {6, 11},
    .disturbance2 = {45, 39}
};

Measurement2 measurement2 = {
    .input2 = true
};

Measurement3 measurement3 = {
    .input3 = 56967,
    .disturbance1 = 35766
};

Measurement4 measurement4 = {
    .output2 = 24
};

Observation1 observation1 = {
    .state1 = 80
};

Observation2 observation2 = {
    .state2 = false
};

Observation3 observation3 = {
    .state3 = 22
};

Target2 target2 = {
    .output2 = 29
};

// ──────────────────────────────────────────────────────────────
// ❷ RULES  (each Prolog clause → one C function)
// ──────────────────────────────────────────────────────────────

/**
 * Implements the two-clause Prolog rule:
 *
 *     measurement10(I,M) :-
 *         measurement1(I,[M1,M2]), M1 < M2, M3 is M2-M1, M is sqrt(M3).
 *
 *     measurement10(I,M1) :-
 *         measurement1(I,[M1,M2]), M1 >= M2.
 *
 * Returns 0.0 and sets `ok` to false if no fact matches.
 */
double measurement10(const char* key, bool* ok) {
    *ok = false;

    if (strcmp(key, "input1") == 0) {
        int m1 = measurement1.input1[0];
        int m2 = measurement1.input1[1];
        *ok = true;

        if (m1 < m2) {
            return sqrt(m2 - m1);
        } else {
            return (double)m1;
        }
    }

    return 0.0;
}

/**
 * Collect every (Actuator, C) pair that satisfies the Prolog query:
 *     ?- control1(_, _).
 */
void control1() {
    bool ok;

    // ── Clause #1  (actuator1) ──────────────────────────────────
    double m1 = measurement10("input1", &ok);

    if (
        ok &&
        measurement2.input2 == true &&
        measurement3.disturbance1 != 0
    ) {
        int d1 = measurement3.disturbance1;

        double c1 = m1 * 19.6;            // proportional part
        double c2 = log10((double)d1);    // compensation part
        double c  = c1 - c2;              // feed-forward control

        printf("actuator1: %.4f\n", c);
    }

    // ── Clause #2  (actuator2) ──────────────────────────────────
    if (
        observation3.state3 &&              // state3 must exist
        measurement4.output2 &&             // output2 must exist
        target2.output2                     // target2 output2 must exist
    ) {
        int p3 = observation3.state3;
        int m4 = measurement4.output2;
        int t2 = target2.output2;

        int e = t2 - m4;                   // error
        if (e != 0) {                      // avoid div-by-zero (Prolog would error out)
            int d = p3 - m4;               // differential error
            double c1 = 5.8 * e;           // proportional part
            double n  = 7.3 / (double)e;   // nonlinear factor
            double c2 = n * d;             // nonlinear differential part
            double c  = c1 + c2;           // P-N-D feedback control

            printf("actuator2: %.4f\n", c);
        }
    }
}

// ──────────────────────────────────────────────────────────────
// ❸ RUN THE SAME QUERY Prolog WOULD
// ──────────────────────────────────────────────────────────────
int main() {
    control1();
    return 0;
}

